File: services/entity-resolution-service/src/fusion_base/domain/deduplication.rs
use serde_json::json;

use crate::models::{cluster::EntityRecord, job::ResolutionJobConfig};

pub fn synthesize_entity_records(
    entity_type: &str,
    config: &ResolutionJobConfig,
) -> Vec<EntityRecord> {
    if entity_type.eq_ignore_ascii_case("organization") {
        synthesize_organization_records(config)
    } else {
        synthesize_person_records(config)
    }
}

fn synthesize_person_records(config: &ResolutionJobConfig) -> Vec<EntityRecord> {
    let profiles = [
        (
            ["John Smith", "Jon Smyth", "J. Smith"],
            [
                "john.smith@acme.com",
                "jon.smyth@acme.com",
                "john.smith+support@acme.com",
            ],
            ["+1 415 555 0100", "+1 (415) 555-0100", "4155550100"],
            "San Francisco",
            "Acme Logistics",
            "100 Market St",
        ),
        (
            ["Ana Perez", "Anna Peres", "A. Perez"],
            [
                "ana.perez@nova.io",
                "anna.peres@nova.io",
                "ana.perez+ops@nova.io",
            ],
            ["+34 91 555 0101", "+34 915550101", "91-555-0101"],
            "Madrid",
            "Nova Energy",
            "Calle Gran Via 44",
        ),
        (
            ["Mark Johnson", "Marc Jonson", "M. Johnson"],
            [
                "mark.johnson@northwind.co",
                "marc.jonson@northwind.co",
                "mark.johnson+help@northwind.co",
            ],
            ["+1 646 555 0102", "+1-646-555-0102", "6465550102"],
            "New York",
            "Northwind Trading",
            "215 Madison Ave",
        ),
        (
            ["Mei Chen", "May Chen", "M. Chen"],
            [
                "mei.chen@harbor.ai",
                "may.chen@harbor.ai",
                "mei.chen+crm@harbor.ai",
            ],
            ["+65 6555 0103", "+65-6555-0103", "65550103"],
            "Singapore",
            "Harbor AI",
            "12 Anson Rd",
        ),
    ];

    build_records(config, &profiles, "person")
}

fn synthesize_organization_records(config: &ResolutionJobConfig) -> Vec<EntityRecord> {
    let profiles = [
        (
            ["Acme Logistics", "ACME Logstics", "Acme Log."],
            [
                "ops@acme-logistics.com",
                "support@acme-logistics.com",
                "hello@acme-logistics.com",
            ],
            ["+1 415 555 1000", "+1 (415) 555-1000", "4155551000"],
            "San Francisco",
            "Acme Logistics",
            "100 Market St",
        ),
        (
            ["Nova Energy", "Nova Energi", "Nova Eng."],
            [
                "info@novaenergy.io",
                "ops@novaenergy.io",
                "hello@novaenergy.io",
            ],
            ["+34 91 555 1001", "+34 915551001", "915551001"],
            "Madrid",
            "Nova Energy",
            "Calle Gran Via 44",
        ),
        (
            ["Northwind Trading", "North Wind Trading", "Northwind Trdg"],
            [
                "ops@northwind.co",
                "support@northwind.co",
                "hello@northwind.co",
            ],
            ["+1 646 555 1002", "+1-646-555-1002", "6465551002"],
            "New York",
            "Northwind Trading",
            "215 Madison Ave",
        ),
    ];

    build_records(config, &profiles, "organization")
}

fn build_records(
    config: &ResolutionJobConfig,
    profiles: &[([&str; 3], [&str; 3], [&str; 3], &str, &str, &str)],
    record_type: &str,
) -> Vec<EntityRecord> {
    let source_labels = if config.source_labels.is_empty() {
        vec!["crm".to_string(), "erp".to_string(), "support".to_string()]
    } else {
        config.source_labels.clone()
    };

    let mut records = Vec::new();
    let target_count = config.record_count.max(9) as usize;

    'outer: for cycle in 0..4 {
        for (profile_index, profile) in profiles.iter().enumerate() {
            for (source_index, source) in source_labels.iter().enumerate() {
                let name = profile.0[source_index % profile.0.len()].to_string();
                let email = profile.1[source_index % profile.1.len()].to_string();
                let phone = profile.2[source_index % profile.2.len()].to_string();
                let external_id = format!("{}-{}-{}", source, profile_index + 1, cycle + 1);

                records.push(EntityRecord {
                    record_id: format!("{source}:{record_type}:{external_id}"),
                    source: source.clone(),
                    external_id: external_id.clone(),
                    display_name: name.clone(),
                    confidence: (0.82 - cycle as f32 * 0.03).clamp(0.45, 0.95),
                    attributes: json!({
                        "name": name,
                        "email": email,
                        "phone": phone,
                        "city": profile.3,
                        "company": profile.4,
                        "address": profile.5,
                        "source_rank": source_index + 1,
                    }),
                });

                if records.len() >= target_count {
                    break 'outer;
                }
            }
        }
    }

    records
}

File: services/entity-resolution-service/internal/domain/deduplication.go
package domain

import (
	"fmt"
	"strings"

	"github.com/openfoundry/openfoundry-go/services/entity-resolution-service/internal/models"
)

// SynthesizeEntityRecords ports `deduplication::synthesize_entity_records`.
func SynthesizeEntityRecords(entityType string, config models.ResolutionJobConfig) []models.EntityRecord {
	if strings.EqualFold(entityType, "organization") {
		return synthesizeOrganizationRecords(config)
	}
	return synthesizePersonRecords(config)
}

type profile struct {
	names    [3]string
	emails   [3]string
	phones   [3]string
	city     string
	company  string
	address  string
}

var personProfiles = []profile{
	{
		names:   [3]string{"John Smith", "Jon Smyth", "J. Smith"},
		emails:  [3]string{"john.smith@acme.com", "jon.smyth@acme.com", "john.smith+support@acme.com"},
		phones:  [3]string{"+1 415 555 0100", "+1 (415) 555-0100", "4155550100"},
		city:    "San Francisco",
		company: "Acme Logistics",
		address: "100 Market St",
	},
	{
		names:   [3]string{"Ana Perez", "Anna Peres", "A. Perez"},
		emails:  [3]string{"ana.perez@nova.io", "anna.peres@nova.io", "ana.perez+ops@nova.io"},
		phones:  [3]string{"+34 91 555 0101", "+34 915550101", "91-555-0101"},
		city:    "Madrid",
		company: "Nova Energy",
		address: "Calle Gran Via 44",
	},
	{
		names:   [3]string{"Mark Johnson", "Marc Jonson", "M. Johnson"},
		emails:  [3]string{"mark.johnson@northwind.co", "marc.jonson@northwind.co", "mark.johnson+help@northwind.co"},
		phones:  [3]string{"+1 646 555 0102", "+1-646-555-0102", "6465550102"},
		city:    "New York",
		company: "Northwind Trading",
		address: "215 Madison Ave",
	},
	{
		names:   [3]string{"Mei Chen", "May Chen", "M. Chen"},
		emails:  [3]string{"mei.chen@harbor.ai", "may.chen@harbor.ai", "mei.chen+crm@harbor.ai"},
		phones:  [3]string{"+65 6555 0103", "+65-6555-0103", "65550103"},
		city:    "Singapore",
		company: "Harbor AI",
		address: "12 Anson Rd",
	},
}

var organizationProfiles = []profile{
	{
		names:   [3]string{"Acme Logistics", "ACME Logstics", "Acme Log."},
		emails:  [3]string{"ops@acme-logistics.com", "support@acme-logistics.com", "hello@acme-logistics.com"},
		phones:  [3]string{"+1 415 555 1000", "+1 (415) 555-1000", "4155551000"},
		city:    "San Francisco",
		company: "Acme Logistics",
		address: "100 Market St",
	},
	{
		names:   [3]string{"Nova Energy", "Nova Energi", "Nova Eng."},
		emails:  [3]string{"info@novaenergy.io", "ops@novaenergy.io", "hello@novaenergy.io"},
		phones:  [3]string{"+34 91 555 1001", "+34 915551001", "915551001"},
		city:    "Madrid",
		company: "Nova Energy",
		address: "Calle Gran Via 44",
	},
	{
		names:   [3]string{"Northwind Trading", "North Wind Trading", "Northwind Trdg"},
		emails:  [3]string{"ops@northwind.co", "support@northwind.co", "hello@northwind.co"},
		phones:  [3]string{"+1 646 555 1002", "+1-646-555-1002", "6465551002"},
		city:    "New York",
		company: "Northwind Trading",
		address: "215 Madison Ave",
	},
}

func synthesizePersonRecords(config models.ResolutionJobConfig) []models.EntityRecord {
	return buildRecords(config, personProfiles, "person")
}

func synthesizeOrganizationRecords(config models.ResolutionJobConfig) []models.EntityRecord {
	return buildRecords(config, organizationProfiles, "organization")
}

func buildRecords(config models.ResolutionJobConfig, profiles []profile, recordType string) []models.EntityRecord {
	sourceLabels := config.SourceLabels
	if len(sourceLabels) == 0 {
		sourceLabels = []string{"crm", "erp", "support"}
	}

	targetCount := int(config.RecordCount)
	if targetCount < 9 {
		targetCount = 9
	}

	records := make([]models.EntityRecord, 0, targetCount)
	for cycle := 0; cycle < 4; cycle++ {
		for profileIndex, p := range profiles {
			for sourceIndex, source := range sourceLabels {
				name := p.names[sourceIndex%len(p.names)]
				email := p.emails[sourceIndex%len(p.emails)]
				phone := p.phones[sourceIndex%len(p.phones)]
				externalID := fmt.Sprintf("%s-%d-%d", source, profileIndex+1, cycle+1)

				confidence := 0.82 - float32(cycle)*0.03
				switch {
				case confidence < 0.45:
					confidence = 0.45
				case confidence > 0.95:
					confidence = 0.95
				}

				records = append(records, models.EntityRecord{
					RecordID:    fmt.Sprintf("%s:%s:%s", source, recordType, externalID),
					Source:      source,
					ExternalID:  externalID,
					DisplayName: name,
					Confidence:  confidence,
					Attributes: map[string]any{
						"name":        name,
						"email":       email,
						"phone":       phone,
						"city":        p.city,
						"company":     p.company,
						"address":     p.address,
						"source_rank": sourceIndex + 1,
					},
				})

				if len(records) >= targetCount {
					return records
				}
			}
		}
	}
	return records
}
